### PR TITLE
Function to pull neso forecast #2

### PR DESCRIPTION
**Pull request by Seao7**
# Pull Request

## Description

This pull request adds a Python script to fetch data from the NESO API. It introduces two main functions:  
1. **API fetch function**: A simple method to retrieve data directly from the NESO API.  
2. **SQL query-based fetch function**: An alternative method to retrieve data using SQL queries.

In addition to the core functionality, this PR includes a test suite with the following test functions:
- **`test_fetch_data_api()`**: Verifies that the `fetch_data` function correctly fetches and processes data from the API, including proper column renaming.
- **`test_fetch_data_sql()`**: Ensures that the `fetch_data_using_sql` function works correctly with SQL queries, processes the data, and renames the columns as expected.
- **`test_data_consistency()`**: Validates that the data fetched from both methods (`fetch_data` and `fetch_data_using_sql`) is consistent.

The code has been formatted using [Black](https://black.readthedocs.io/en/stable/) to ensure consistency and adhere to coding style guidelines.

### Motivation

This script simplifies the process of fetching data from the NESO API, making it easier for users to retrieve and process data in two ways: directly via API calls or through SQL queries. The flexibility of these methods allows for use in a variety of data-fetching scenarios, facilitating analysis and further processing.

### Dependencies

- `pytest`: For testing the functionality of the fetch methods.
- No additional dependencies beyond the standard Python libraries (`urllib`, `json`, and `pandas`) are required.

## How Has This Been Tested?

The functionality of the `fetch_data` and `fetch_data_using_sql` functions has been validated through the following test cases:
1. **`test_fetch_data_api()`**: Ensures correct functionality of the API fetch method.
2. **`test_fetch_data_sql()`**: Confirms the proper operation of the SQL fetch method.
3. **`test_data_consistency()`**: Verifies that the data fetched by both methods is consistent in terms of structure and content.

To reproduce the tests:
- Run the test functions provided in the script.
- Ensure internet connectivity and valid access to the NESO API.

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md).  
- [x] I have performed a self-review of my own code.  
- [ ] I have updated the documentation to include details about the new script.  
- [x] I have added tests that validate the correctness of my code.  
- [x] I have checked my code and corrected any misspellings.

### CODE REVIEW

File: fetch_data.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        df.rename(columns=rename_columns, inplace=True)
+
+        return df
+
+    except Exception as e:
+        print(f"An error occurred: {e}")
+        return pd.DataFrame()
+
+
+def test_fetch_functions():
```
Comment (peterdudfield) [inline fetch_data.py:104]: COuld you put this in a new file, perhaps `tests/test_fetch_data.py`?

File: fetch_data.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+import urllib.request
+import urllib.parse
+import json
+import pandas as pd
+
+
+def fetch_data(
+    resource_id: str, limit: int, columns: list, rename_columns: dict
+) -> pd.DataFrame:
+    """
```
Comment (peterdudfield) [inline fetch_data.py:14]: nice descirptions and type hinting

File: fetch_data.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+    print("Testing fetch_data function...")
+    df_api = fetch_data(resource_id, limit, columns, rename_columns)
+    assert not df_api.empty, "fetch_data returned an empty DataFrame!"
+    assert set(df_api.columns) == set(
+        rename_columns.values()
+    ), "Column names do not match after renaming!"
+    print("fetch_data passed the tests.")
+
+    print("Testing fetch_data_using_sql function...")
```
Comment (peterdudfield) [inline fetch_data.py:129]: can you split these into 2 separate tests?

**Comment by Seao7** [comment]
Alright. I'll create a separate test file with two separate tests.

**Comment by Seao7** [comment]
@peterdudfield I have made a separate tests folder and am now using pytest to test the functions. I hope that is fine. Have converted the single test function into 3 separate functions as well.

File: tests/test_fetch_data.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    Validate that the data fetched by fetch_data and fetch_data_using_sql are consistent.
+    """
+    df_api = fetch_data(resource_id, limit, columns, rename_columns)
+    df_sql = fetch_data_using_sql(sql_query, columns, rename_columns)
+    assert df_api.equals(
+        df_sql
+    ), "Data from fetch_data and fetch_data_using_sql are inconsistent!"
+
+
+if __name__ == "__main__":
```
Comment (peterdudfield) [inline tests/test_fetch_data.py:77]: this can be removed

Comment (Seao7) [reply tests/test_fetch_data.py]: @peterdudfield Have removed the __main__ call.

**Review by peterdudfield — APPROVED** [review]

**Comment by Seao7** [comment]
@peterdudfield Hey, should I start working on #4? Is this issue resolved fully or need some other improvements?

**Comment by peterdudfield** [comment]
@all-contributors please add @peterdudfield  for ideas

**Pull request merged**